generators/modelIntensity.cc

```diff
@@ -0,0 +1,210 @@
+		const std::string& waveName = _fitResult->waveName(wave);
+		// no amplitude for 'flat' wave
+		if (waveName == "flat") {
+			if (_amplitudes[wave]) {
+				printErr << "amplitude for flat wave was added. Aborting..." << std::endl;
+				return false;
+			}
+			continue;
+		}
+		if (not _amplitudes[wave]) {
+			printErr << "amplitude for wave '" << waveName << "' was not added. Aborting..." << std::endl;
+			return false;
+		}
+
+		const isobarDecayTopologyPtr decay  = _amplitudes[wave]->decayTopology();
+		const isobarDecayVertexPtr   vertex = decay->XIsobarDecayVertex();
+		prodKinParticleNames[0] = vertex->parent()->name();
+		break;
+	}
+
+	return initAmplitudes(prodKinParticleNames, decayKinParticleNames, true);
+}
+
+
+bool
+rpwa::modelIntensity::initAmplitudes(const std::vector<std::string>& prodKinParticleNames,
+                                     const std::vector<std::string>& decayKinParticleNames,
+                                     const bool                      fromXDecay)
+{
+	_amplitudesInitialized = false;
+	_amplitudesFromXDecay  = fromXDecay;
+
+	for (size_t wave = 0; wave < _fitResult->nmbWaves(); ++wave) {
+		const std::string& waveName = _fitResult->waveName(wave);
+		// no amplitude for 'flat' wave
+		if (waveName == "flat") {
+			if (_amplitudes[wave]) {
+				printErr << "amplitude for flat wave was added. Aborting..." << std::endl;
+				return false;
+			}
+			continue;
+		}
+		if (not _amplitudes[wave]) {
+			printErr << "amplitude for wave '" << waveName << "' was not added. Aborting..." << std::endl;
+			return false;
+		}
+
+		// start amplitude from X decay
+		if (_amplitudesFromXDecay) {
+			const isobarDecayTopologyPtr decay  = _amplitudes[wave]->decayTopology();
+			const isobarDecayVertexPtr   vertex = decay->XIsobarDecayVertex();
+			_amplitudes[wave]->setDecayTopology(rpwa::createIsobarDecayTopology(decay->subDecayConsistent(vertex)));
+		}
+		_amplitudes[wave]->init();
+		if (not _amplitudes[wave]->decayTopology()->initKinematicsData(prodKinParticleNames, decayKinParticleNames)) {
+			printErr << "could not initialize kinematics data for amplitude of wave '" << waveName << "'. Aborting..." << std::endl;
+			return false;
+		}
+	}
+
+	_amplitudesInitialized = true;
+	return true;
+}
+
+
+double
+rpwa::modelIntensity::getIntensity(const std::vector<unsigned int>& waveIndices,
+                                   const std::vector<TVector3>&     prodKinMomenta,
+                                   const std::vector<TVector3>&     decayKinMomenta) const
+{
+	const std::vector<std::complex<double> > amplitudes = getAmplitudes(prodKinMomenta, decayKinMomenta);
+
+	double intensity = 0;
+	for (std::set<int>::const_iterator it=_allRefls.begin(); it!=_allRefls.end(); ++it) {
+		std::complex<double> amp = 0;
+		for (size_t i=0; i<waveIndices.size(); ++i) {
+			const unsigned int waveIndex = waveIndices[i];
+			if (_refls[waveIndex] == *it) {
+				amp += _fitResult->prodAmp(waveIndex) * amplitudes[waveIndex];
```

Comment:
possible optimization would be to store decay amplitudes in 2D structure [refl][wave_index(refl)]
not sure though, whether this would be worth the effort